### ISSUE TITLE
[bug]: Infinite loop of "Indicated Video Update", max CPU usage

### ISSUE THREAD

**Issue by duncannz** [asterisk/asterisk#780]
### Severity

Major

### Versions

20.6.0,20.8.1

### Components/Modules

Local,PJSIP

### Operating Environment

Debian 11.7

### Frequency of Occurrence

Constant

### Issue Description

I am experiencing:

* Extremely high / maxed out CPU usage
* Log messages such as `Exceptionally long queue length (0 voice / 150 total) queuing to Recorder/ARI-00000001;1` are logged repeatedly for as long as the bridge exists
* If I turn on debug logging, the messages
  ```
  [2024-06-25 01:14:04] DEBUG[3144][C-00000003]: chan_pjsip.c:1644 chan_pjsip_indicate:  PJSIP/webrtc_client-00000001: Indicated Video Update
  [2024-06-25 01:14:04] DEBUG[3144][C-00000003]: chan_pjsip.c:1875 chan_pjsip_indicate:  PJSIP/webrtc_client-00000001
  ```
  are logged repeatedly at an extremely high rate for as long as the bridges exist

Bridge 1:

```
Id: 01J16F9GK7X976AJQ6F596SKN6
Type: stasis
Technology: softmix
Subclass: stasis
Creator: Stasis
Name:
Video-Mode: talker
Video-Source-Id:
Num-Channels: 3
Num-Active: 3
Duration: 00:00:30
Channel: PJSIP/webrtc_client-00000000
Channel: Recorder/ARI-00000001;2
Channel: Local/+XXXXX@reflector-00000000;1
```

Bridge 2:

```
Id: 01J16F9GT51PEHR8F66WGVE7D6
Type: stasis
Technology: softmix
Subclass: stasis
Creator: Stasis
Name:
Video-Mode: talker
Video-Source-Id:
Num-Channels: 3
Num-Active: 3
Duration: 00:00:38
Channel: PJSIP/webrtc_client-00000001
Channel: Recorder/ARI-00000002;2
Channel: Local/+XXXXX@reflector-00000000;2
```

This issue happens on 100% of calls that have the above configuration (one PJSIP channel in each bridge, one local channel connecting the two bridges, and each bridge being recorded).

The only other thing I could find on Google for "Indicated Video Update" is [this post](https://asteriskfaqs.org/2021/05/05/asterisk-users/cpu-spike.html), which describes the exact same issue, but in Asterisk 16.17.0, so I suspect this issue has been around for some time.

I am not using video in any way, and am just creating the bridges with a single argument `type: "mixing"` via ARI. The PJSIP channels are using Opus over WebRTC.

If I have missed any details that would be required to reproduce the issue, please let me know.

### Asterisk Issue Guidelines

- [X] Yes, I have read the Asterisk Issue Guidelines

**Comment by duncannz** [comment]
I have tested adding the argument `video_single` when creating the bridge (so that `Video-Mode` is `single`), but the issue still occurs.

I have also tried applying the following patch and adding the argument `video_none` when creating the bridge instead (so that `Video-Mode` is `none`), but even then, the issue still occurs.

```
--- res/res_stasis.c
+++ res/res_stasis.c
@@ -822,6 +822,8 @@ static struct ast_bridge *bridge_create_common(const char *type, const char *nam
 			video_mode = AST_BRIDGE_VIDEO_MODE_SFU;
 		} else if (!strcmp(requested_type, "video_single")) {
 			video_mode = AST_BRIDGE_VIDEO_MODE_SINGLE_SRC;
+		} else if (!strcmp(requested_type, "video_none")) {
+			video_mode = AST_BRIDGE_VIDEO_MODE_NONE;
 		} else if (!strcmp(requested_type, "sdp_label")) {
 			send_sdp_label = 1;
 		}

--- res/stasis/stasis_bridge.c
+++ res/stasis/stasis_bridge.c
@@ -313,7 +313,7 @@ struct ast_bridge *bridge_stasis_new(uint32_t capabilities, unsigned int flags,
 		ast_bridge_set_video_update_discard(bridge, 5);
 	} else if (video_mode == AST_BRIDGE_VIDEO_MODE_SINGLE_SRC) {
 		ast_bridge_set_single_src_video_mode(bridge, NULL);
-	} else {
+	} else if (video_mode == AST_BRIDGE_VIDEO_MODE_TALKER_SRC) {
 		ast_bridge_set_talker_src_video_mode(bridge);
 	}
```

**Comment by duncannz** [comment]
Surprisingly, the `video_sfu` type does avoid the issue. (And it's not just because of the following lines at `res_stasis.c:830`, because even removing the condition so that they also run for the other video modes doesn't help.)

```
/* For an SFU video bridge we ensure it always remains in multimix for the best experience. */
if (video_mode == AST_BRIDGE_VIDEO_MODE_SFU) {
	capabilities = AST_BRIDGE_CAPABILITY_MULTIMIX;
	flags &= ~AST_BRIDGE_FLAG_SMART;
}
```

Unfortunately though, in `video_sfu` mode, we face another issue where if the local channel is answered quickly, the bridge audio gets completely broken for the rest of its lifetime, changing to just a ticking sound that vaguely corresponds to actual audio, but I guess that's a separate issue that I don't quite have enough info to report.

<img width="1060" alt="Screenshot 2024-06-25 at 2 06 58 PM" src="https://github.com/asterisk/asterisk/assets/68309395/b646fa58-ffae-47e1-9d9d-44274c716f59">

**Comment by gtjoseph** [comment]
@duncannz Can you give us the exact ARI calls you're using to set the call up? I.E.  Creating the bridge, adding channels, starting recording, etc.

If you don't start recording, do you still get the flood of messages?
How about if you only record one of the bridges and not the other?
Which WebRTC client are you using?
Are you seeing SIP messages during the flood like reINVITEs or UPDATES?

**Comment by duncannz** [comment]
@gtjoseph Please see attached log of the ARI calls and a summary below:

[ari_logs.txt](https://github.com/user-attachments/files/15981459/ari_logs.txt)

On the first side:

1. `webrtc_client` originates a channel (`PJSIP_CHANNEL_A`)
2. ARI answers this channel, creates bridge `BRIDGE_A`, and adds `PJSIP_CHANNEL_A` to `BRIDGE_A`
3. ARI originates a call to `Local/+XXXXX@reflector/n` (channel `LOCAL_CHANNEL`)
4. After `LOCAL_CHANNEL` gets answered, it also adds it to `BRIDGE_A`

Now on the other side:

1. ARI creates bridge `BRIDGE_B` and adds the other side of the local channel to it (`LOCAL_CHANNEL;2`)
2. `webrtc_client` originates a channel (`PJSIP_CHANNEL_B`) (in order to join the call)
3. ARI answers both this channel and the other side of the local channel (`LOCAL_CHANNEL;2`) and adds them both to `BRIDGE_B`

Then we set up recordings for both bridges. To answer your question above, I have found that the issue only occurs after setting up recordings for both bridges. Recording just one bridge or neither avoids the issue.

The WebRTC client is a web browser e.g. Google Chrome, but it communicates with Asterisk via SIP so as far as I know it could be replaced with a normal SIP client for the purpose of this issue.

There are no SIP messages during the flood or any SIP messages other than normal for setting up and eventually tearing down the PJSIP channels.

![Screenshot 2024-06-26 at 12 17 27 PM](https://github.com/asterisk/asterisk/assets/68309395/e87a8d9b-df53-43e4-953c-00ce2c2e9925)

**Comment by gtjoseph** [comment]
Thanks for the additional info.  We have an internal issue open for this that's at the top of the backlog.  It's not going to be a trivial fix however so it may be a while.

**Comment by duncannz** [comment]
Thanks very much @gtjoseph!

I'm planning on working around this for now by using a PJSIP channel pointed at localhost instead of a Local channel. This seems to avoid the issue, but can you confirm whether this should fully avoid the issue or might just appear to be avoiding it due to different timing?

In other words, is your understanding that the issue is specific to the Local channel driver?

**Comment by gtjoseph** [comment]
We're not actually sure if the issue is local channel specific.  It just so happens that every instance of this issue we've seen involves multiple bridges and local channels are the normal way to connect the two together.

In looking at your ari log I do have some questions and suggestions for things you could try...

 * What's the time delay between when you get the call from PJSIP_CHANNEL_A and the call from PJSIP_CHANNEL_B?
 * In one reported use case, PJSIP_CHANNEL_A was an agent that stayed up and serviced multiple sequential PJSIP_CHANNEL_B callers.  Is that the case here or do both PJSIP channels get torn down after each call?
 * Unless you specifically need to prevent early media on PJSIP_CHANNEL_A, don't call answer on it. You should only call answer on LOCAL_CHANNEL;2 and PJSIP_CHANNEL_B.
 * Try calling dial on LOCAL_CHANNEL before you put it into BRIDGE_A.
 * Since you've already answered PJSIP_CHANNEL_A, do you really need to call ring on LOCAL_CHANNEL;2?  
 * It looks like you're calling play on PJSIP_CHANNEL_A right after you add it to BRIDGE_A but you're not stopping it until after you've started recording on BRIDGE_A.  Can you try stopping it before you answer PJSIP_CHANNEL_B?

**Comment by gtjoseph** [comment]
Thanks for the info.

> * Try calling dial on LOCAL_CHANNEL before you put it into BRIDGE_A.
Sorry just to clarify, does "it" refer to LOCAL_CHANNEL or PJSIP_CHANNEL_A? Are you suggesting changing the sequence of existing ARI calls or adding another?

Sorry, I meant adding a call to /channels/LOCAL_CHANNEL/dial just before you call /bridges/BRIDGE_A/addChannel to add LOCAL_CHANNEL to BRIDGE_A.

**Comment by duncannz** [comment]
I have just tried that but then I get an error from `/channels/LOCAL_CHANNEL/dial`:

```
{"message":"Channel is not in the 'Down' state"}
```

but despite the error, it still seems to have an effect as I keep hearing ringback rather than the call audio on the caller side (and no audio on the recipient side). It does also avoid the original issue, but with this obvious significant caveat. I'm not exactly sure what effect it was supposed to have, though.

**Comment by gtjoseph** [comment]
@duncannz I'm not able to reproduce the issue just using the ari log you provided earlier.  Can you post the dialplan that goes with your scenario?  Specifically, what's in the "reflector" context?  When the PJSIP channels call in, is there anything other than a call to "Stasis()" being done?

What codec is being used by the two PJSIP channels?  Are they using the same codec or different?  If different, what happens if you force them to use the same codec?

It doesn't look like you're setting the "formats" parameter on the call that creates the local channels.  What happens if you explicitly set it to the same codec used by the two PJSIP channels?

**Comment by gtjoseph** [comment]
Also...  What platform is your ARI app?  Node-JS, Python, etc?  What ARI client library are you using, if any?

**Comment by gtjoseph** [comment]
OK, got it figured out.  @duncannz can you test the fix in PR #810 ?  It should apply cleanly to any active Asterisk branch.

**Comment by duncannz** [comment]
I can confirm this fix resolves the issue for me. Thank you so much!
